group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java

```diff
@@ -0,0 +1,581 @@
+                    .withPartitionWriter(writer)
+                    .withLoader(loader)
+                    .withCoordinatorBuilderSupplier(supplier)
+                    .build();
+
+            return new GroupCoordinatorService(
+                logContext,
+                config,
+                runtime
+            );
+        }
+    }
+
+    /**
+     * The logger.
+     */
+    private final Logger log;
+
+    /**
+     * The group coordinator configurations.
+     */
+    private final GroupCoordinatorConfig config;
+
+    /**
+     * The coordinator runtime.
+     */
+    private final CoordinatorRuntime<ReplicatedGroupCoordinator, Record> runtime;
+
+    /**
+     * Boolean indicating whether the coordinator is active or not.
+     */
+    private final AtomicBoolean isActive = new AtomicBoolean(false);
+
+    /**
+     * The number of partitions of the __consumer_offsets topics. This is provided
+     * when the component is started.
+     */
+    private volatile int numPartitions = -1;
+
+    /**
+     *
+     * @param logContext
+     * @param config
+     * @param runtime
+     */
+    GroupCoordinatorService(
+        LogContext logContext,
+        GroupCoordinatorConfig config,
+        CoordinatorRuntime<ReplicatedGroupCoordinator, Record> runtime
+    ) {
+        this.log = logContext.logger(CoordinatorLoader.class);
+        this.config = config;
+        this.runtime = runtime;
+    }
+
+    /**
+     * Throws CoordinatorNotAvailableException if the not active.
+     */
+    private void throwIfNotActive() {
+        if (!isActive.get()) {
+            throw Errors.COORDINATOR_NOT_AVAILABLE.exception();
+        }
+    }
+
+    /**
+     * @return TODO
+     */
+    private TopicPartition topicPartitionFor(
+        String groupId
+    ) {
+        return new TopicPartition(Topic.GROUP_METADATA_TOPIC_NAME, partitionFor(groupId));
+    }
+
+    /**
+     * See {@link GroupCoordinator#partitionFor(String)}
+     */
+    @Override
+    public int partitionFor(
+        String groupId
+    ) {
+        throwIfNotActive();
+        return Utils.abs(groupId.hashCode()) % numPartitions;
+    }
+
+    /**
+     * See {@link GroupCoordinator#consumerGroupHeartbeat(RequestContext, ConsumerGroupHeartbeatRequestData)}.
+     */
+    @Override
+    public CompletableFuture<ConsumerGroupHeartbeatResponseData> consumerGroupHeartbeat(
+        RequestContext context,
+        ConsumerGroupHeartbeatRequestData request
+    ) {
+        if (!isActive.get()) {
+            return FutureUtils.failedFuture(Errors.COORDINATOR_NOT_AVAILABLE.exception());
+        }
+
+        return runtime.scheduleWriteOperation(
+            "consumer-group-heartbeat",
+            topicPartitionFor(request.groupId()),
+            coordinator -> coordinator.consumerGroupHeartbeat(context, request)
+        ).exceptionally(exception -> {
+            if (exception instanceof UnknownTopicOrPartitionException) {
```

Comment:
have we always converted these errors as such? I see in GroupMetadataManager things are slightly different.

Reply:
Yes. We have similar logic in the scala code [here](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala#L288). The main difference is that we have to do it at a different place now.

Reply:
I saw this code, but it seems like we handle it different right?

notenoughreplicas moved to not coordinator when it used to be coordinator not available for example.

Reply:
That seems to be a mistake. I will check this tomorrow.